Source/BSGOutOfMemoryWatchdog.m

```diff
@@ -0,0 +1,237 @@
+    }
+    self.watching = NO;
+    [self deleteSentinelFile];
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+    @try {
+        [[Bugsnag configuration]
+            removeObserver:self
+                forKeyPath:NSStringFromSelector(@selector(releaseStage))];
+    } @catch (NSException *exception) {
+        // Shouldn't happen, but if for some reason, unregistration happens
+        // without registration, catch the resulting exception.
+    }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSString *, id> *)change
+                       context:(void *)context {
+    self.cachedFileInfo[@"app"][@"releaseStage"] = change[NSKeyValueChangeNewKey];
+    [self writeSentinelFile];
+}
+- (void)handleTransitionToForeground:(NSNotification *)note {
+    self.cachedFileInfo[@"app"][@"inForeground"] = @YES;
+    [self writeSentinelFile];
+}
+
+- (void)handleTransitionToBackground:(NSNotification *)note {
+    self.cachedFileInfo[@"app"][@"inForeground"] = @NO;
+    [self writeSentinelFile];
+}
+
+- (void)handleLowMemoryChange:(NSNotification *)note {
+    self.cachedFileInfo[@"device"][@"lowMemory"] = [NSDate date];
+    [self writeSentinelFile];
+}
+
+- (void)handleUpdateSession:(NSNotification *)note {
+    id session = [note object];
+    NSMutableDictionary *cache = (id)self.cachedFileInfo;
+    if (session) {
+        cache[@"session"] = session;
+    } else {
+        [cache removeObjectForKey:@"session"];
+    }
+    [self writeSentinelFile];
+}
+
+- (BOOL)computeDidOOMLastLaunchWithConfig:(BugsnagConfiguration *)config {
+    if ([[NSFileManager defaultManager] fileExistsAtPath:self.sentinelFilePath]) {
+        NSDictionary *lastBootInfo = [self readSentinelFile];
+        if (lastBootInfo != nil) {
+            self.lastBootCachedFileInfo = lastBootInfo;
+            NSString *lastBootAppVersion =
+                [lastBootInfo valueForKeyPath:@"app.version"];
+            NSString *lastBootOSVersion =
+                [lastBootInfo valueForKeyPath:@"device.osBuild"];
+            BOOL lastBootInForeground =
+                [[lastBootInfo valueForKeyPath:@"app.inForeground"] boolValue];
+            NSString *osVersion = [BSG_KSSystemInfo osBuildVersion];
+            NSDictionary *appInfo = [[NSBundle mainBundle] infoDictionary];
+            NSString *appVersion =
+                [appInfo valueForKey:(__bridge NSString *)kCFBundleVersionKey];
+            BOOL sameVersions = [lastBootOSVersion isEqualToString:osVersion] &&
+                                [lastBootAppVersion isEqualToString:appVersion];
+            BOOL shouldReport = config.reportBackgroundOOMs || lastBootInForeground;
+            [self deleteSentinelFile];
+            return sameVersions && shouldReport;
+        }
+    }
+    return NO;
+}
+
+- (void)deleteSentinelFile {
+    [[NSFileManager defaultManager] removeItemAtPath:self.sentinelFilePath
```

Comment:
Should we fall back to using `unlink` here if this fails like we do for `didCrashSentinelPath`?